plugins/base/src/main/kotlin/renderers/html/HtmlTemplater.kt

```diff
@@ -0,0 +1,78 @@
+package org.jetbrains.dokka.base.renderers.html
+
+import freemarker.cache.ClassTemplateLoader
+import freemarker.cache.FileTemplateLoader
+import freemarker.cache.MultiTemplateLoader
+import freemarker.template.Configuration
+import freemarker.template.Template
+import freemarker.template.TemplateExceptionHandler
+import kotlinx.coroutines.async
+import kotlinx.coroutines.runBlocking
+import org.jetbrains.dokka.base.DokkaBase
+import org.jetbrains.dokka.base.DokkaBaseConfiguration
+import org.jetbrains.dokka.plugability.DokkaContext
+import org.jetbrains.dokka.plugability.configuration
+import java.io.StringWriter
+
+
+enum class DokkaTemplateTypes(val path: String) {
+    BASE("base.ftl")
+}
+
+typealias TemplateMap = Map<String, Any?>
+
+class HtmlTemplater(
+    context: DokkaContext
+) {
+    private val configuration = configuration<DokkaBase, DokkaBaseConfiguration>(context)
+    private val templaterConfiguration =
+        Configuration(Configuration.VERSION_2_3_31).apply { configureTemplateEngine() }
+
+    private fun Configuration.configureTemplateEngine() {
+        val loaderFromResources = ClassTemplateLoader(javaClass, "/dokka/templates")
+        templateLoader = configuration?.templatesDir?.let {
+            MultiTemplateLoader(
+                arrayOf(
+                    FileTemplateLoader(it),
+                    loaderFromResources
+                )
+            )
+        } ?: loaderFromResources
+
+        unsetLocale()
+        defaultEncoding = "UTF-8"
+        templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER
+        logTemplateExceptions = false
+        wrapUncheckedExceptions = true
+        fallbackOnNullLoopVariable = false
+    }
+
+    fun setupSharedModel(model: TemplateMap) {
+        templaterConfiguration.setSharedVariables(model)
+    }
+
+    private val cachedTemplates: MutableMap<DokkaTemplateTypes, Template> = mutableMapOf()
+
+    fun renderFromTemplate(
+        templateType: DokkaTemplateTypes,
+        generateModel: () -> TemplateMap
+    ): String {
+        val cachedTemplate = cachedTemplates[templateType]
+        val out = StringWriter()
+        if (cachedTemplate == null) {
+            runBlocking {
+                val templateDeferred = async { templaterConfiguration.getTemplate(templateType.path) }
```

Comment:
Yeah, but since you call `getTemplate`, it will be stored in that cache regardless, and it'll perform the same check every 5 seconds, no?

If none of the checks are causing problems, I don't see why they should be avoided. Moreover, as a plugin developer, if I saw that FreeMarker was used for templating, I'd expect to be able to use all its features and be able to configure it the way I want (to a sensible extent)

Right now I have following concerns:
* Current implementation with `cachedTemplates` is not thread safe (and simply making it `ConcurrentHashMap` will not solve it completely)
* `cachedTemplates` is inconsistent with FreeMarker's internal cache, which can still be used in other places. Memory leaks may occur
* There's no invalidation and no easy way to add any. This isn't such a big deal now, but it's nice to have some out of the box

Unless there are specific problems that `cachedTemplates` solves or some benchmarks that indicate that we need an **additional** cache, I don't see a reason why we need it

Reply:
It is my bad, It seems I have forgotten about thread safety in coroutine context.